Add Pytorch domain specifc ONNX schema for SparseNN ops

yinghai opened the pull request:
Summary: as the tile said

Differential Revision: D10204691

bddppq commented on caffe2/operators/fully_connected_op.cc:
```diff
@@ -154,7 +154,8 @@ OPERATOR_SCHEMA(FCTransposed)
 FCTransposed stands for calling blass with no noTrans, noTrans
-)DOC");
+)DOC")
+    .InheritOnnxSchema("FCTransposed");
```
Maybe we can add an overload for `InheritOnnxSchema` that doesn't need to pass the same caffe2 schema name:
```c++
OpSchema& OpSchema::InheritOnnxSchema() {
  return InheritOnnxSchema(type_);
}
```

yinghai replied:
Cool! I would like to do it in a separate diff as it will touch `core` and many other files.

bddppq replied:
sure

bddppq commented on caffe2/onnx/torch_ops/defs.cc:
```diff
@@ -5,17 +5,103 @@
 
 ONNX_PYTORCH_OPERATOR_SET_SCHEMA(
-    DUMMY_TEST_ONLY,
+    ExpandDims,
```
Looks like this is equivalent to onnx's [Unsqueeze](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Unsqueeze)

bddppq commented on caffe2/onnx/torch_ops/defs.cc:
```diff
@@ -5,17 +5,103 @@
+            "Constrain input and output types to float tensors."));
+
+ONNX_PYTORCH_OPERATOR_SET_SCHEMA(
+    BatchMatMul,
```
This should be equivalent to ONNX's MatMul

yinghai replied:
ONNX matmul supports batch?

bddppq replied:
Yep it behaves like [numpy's matmul](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html).

yinghai replied:
Could we keep it as BatchMatMul, as Glow is implementing BatchMatMul and right now, for these ops I would like a 1-1 mapping between Glow and C2 to help with the debug. We can further reduce it into other primitives once we have a baseline.

bddppq replied:
The onnx -> caffe2 backend actually translates ONNX's MatMul to c2's BatchMatMul already, so you probably don't need this?

Commit:
Add Pytorch domain specifc ONNX schema for SparseNN ops (#12338)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12338

as the tile said

Differential Revision: D10204691

fbshipit-source-id: 847c845007ce26f7b7b4117b5c311f9b9ec4c513

yinghai commented:
Test errors on CUDA platforms are NCCL related and should not be related to this PR.
